### PR TITLE
Sprint 08

### PR DESCRIPTION
**Pull request by DanielCalimayor**

### CODE REVIEW

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
[link text](https://)# Daniel Calimayor- Permutation_Importances



# Part 1. Cleaning data/Imports

***
**What went well:**

-Nice function to take in data with multiple datasets using glob:

```python
from glob import glob
import pandas as pd

def search_column(name):
    for path in glob('competition_data/*.csv'):
        df = pd.read_csv(path)
        if name in df.columns:
            print(path, df.shape)
            print(df.columns.tolist(), '\n')
search_column('thickness')

```


-Clean and intuitive code to clean data:

```python

mport pandas as pd
mats = pd.read_csv('competition_data/bill_of_materials.csv')

ac = mats.melt(id_vars='tube_assembly_id', 
                                     value_vars=[f'component_id_{n}' for n in range(1,9)])

ac = (ac
                       .sort_values(by='tube_assembly_id')
                       .dropna()
                       .rename(columns={'value': 'component_id'}))

ac.head(10)


```


## XGBoost

- Great job bringing in XGBoost and making it work out:
```python

import category_encoders as ce
from xgboost import XGBRegressor
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

encoder = ce.OrdinalEncoder()
X_train_encoded = encoder.fit_transform(X_train)
X_val_encoded = encoder.transform(X_val)
eval_set = [(X_train_encoded, y_train_log), 
            (X_val_encoded, y_val_log)]
model = XGBRegressor(n_estimators=2000, n_jobs=-1)
model.fit(X_train_encoded, y_train_log, 
          eval_set=eval_set, eval_metric='rmse', early_stopping_rounds=50)



```

## Decision Tree 


- Good job at using sklearn and DecisionTreeClassifier. This is commonly used in the data science industry and you should remmber this:

```python
# Documentation 
# https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier(random_state=20)
dt.fit(train_location, y_train)
print('Decision Tree')
print('Train Accuracy', dt.score(train_location, y_train))
print('Validation Accuracy', dt.score(val_location, y_val))
```



## Plotting features



**What went well:**

- Great job importing pdpbox.pdp and plotting the grid type of the two features :

```python
from pdpbox.pdp import pdp_interact, pdp_interact_plot

features = ['min_order_quantity', 'quantity']

interaction = pdp_interact(
    model=model, 
    dataset=X_val_encoded, 
    model_features=X_val_encoded.columns, 
    features=features
)

pdp_interact_plot(interaction, plot_type='grid', feature_names=features);

```


** What could be improved: 

- Could see higher accuracy in the end, and past assignments should be committed as well 

#Stretch Goal (PARTIAL)



# What to work on
- [x] Comment more in code 
- [x] Make more PR's for assignments!!
- [x] Come to PM for questions or help

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
Mispush

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# Daniel C.
## Sprint Challenge: Regression 2 


# Part 1. Wrangle Data

***
**What went well:**

-Nice merge using pandas and .merge:

```python
train = (train.merge(store, how='left')).fillna(0)
test = (test.merge(store, how = 'left')).fillna(0)


```
**What could be better**


-You could be using .describe to obtain information from the data. Other use case is making sure to exclude the number in trainstore: 

```python 
train_store.describe()
train_store.describe(exclude='number')


```

- Plot could be easily applied using seaborn:

```python 
%matplotlib inline
import seaborn as sns
sns.distplot(y_train);


sns.distplot(y_train_log);

```



# Fit and validate your model

- Good cross val score using xgboost and category encoding: 

```python

imp = SimpleImputer(missing_values=np.nan, strategy='mean')
encoder = ce.OrdinalEncoder()
X_train_encoded = encoder.fit_transform(X_train)
X_val_encoded = encoder.transform(X_val)
X_test_encoded = encoder.fit_transform(X_test)
X_train_encoded = imp.fit_transform(X_train_encoded)
X_val_encoded = imp.transform(X_val_encoded)
X_test_encoded = imp.transform(X_test_encoded)

eval_set = [(X_train_encoded, y_train_log), 
            (X_val_encoded, y_val_log)]
model = XGBRegressor(n_estimators=2000, n_jobs=-1)
model.fit(X_train_encoded, y_train_log, 
          eval_set=eval_set, eval_metric='rmse', early_stopping_rounds=50)

```

**What could be better**

- Making a pipeline with k_means and XGBRegressor is an ideal solution as well (slightly better score):

```python
encoder = ce.OrdinalEncoder()
x_train_encoded = encoder.fit_transform(x_train)
x_val_encoded = encoder.transform(x_val)
eval_set = [(x_train_encoded, y_train_log), 
            (x_val_encoded, y_val_log)]
model = XGBRegressor(n_estimators=2000, n_jobs=-1)
model.fit(X_train_encoded, y_train_log, 
          eval_set=eval_set, eval_metric='rmse', early_stopping_rounds=50)

```


# Plot model interpretation visualizations

- Great try! Ideal solution to this visualization is using pdp :

```python
from pdpbox.pdp import pdp_isolate, pdp_plot

feature = 'Assortment'

isolated = pdp_isolate(
    model=model, 
    dataset=x_val_encoded, 
    model_features=x_val_encoded.columns, 
    feature=feature
)

pdp_plot(isolated, feature_name=feature);


#conf matrix


from pdpbox.pdp import pdp_interact, pdp_interact_plot

features = ['Promo2SinceWeek', 'StoreType']

interaction = pdp_interact(
    model=model, 
    dataset=x_val_encoded, 
    model_features=x_val_encoded.columns, 
    features=features
)

pdp_interact_plot(interaction, plot_type='grid', feature_names=features);
```
**What could be better**

- pdp is a great way to see optimal category, but it's better to use it to isolate the model and columns for a cleaner code solution:

```python
from pdpbox.pdp import pdp_isolate, pdp_plot

feature = 'Customers'

isolated = pdp_isolate(
    model=model, 
    dataset=X_val_encoded, 
    model_features=X_val_encoded.columns, 
    feature=feature, 
    num_grid_points=20 
)

pdp_plot(isolated, feature_name=feature);
```


- Another example:

```python
from pdpbox.pdp import pdp_interact, pdp_interact_plot

features = ['Customers', 'CompetitionDistance']

interaction = pdp_interact(
    model=model, 
    dataset=X_val_encoded, 
    model_features=X_val_encoded.columns, 
    features=features
)

pdp_interact_plot(interaction, plot_type='grid', feature_names=features);
```


# What to work on
- [x] Try to read code backwards (makes for less complicated coding)
- [x] Come to PM for questions or help
